### ISSUE TITLE
Wayland, MAME, BGFX and SDL_GetWindowWMInfo()

### ISSUE THREAD

**Issue by sjnewbury** [libsdl-org/SDL#7835]
MAME has pretty much given up on getting Wayland support working with the BGFX renderer, so I decided to take a look at why it doesn't work.

https://github.com/mamedev/mame/issues/10859#issuecomment-1413254155

After fixing up the obvious things to make it compile, I discovered the root of the problem is BGFX fails to create a Surface on the Wayland window because the egl_window handle is invalid.

It seems SDL_GetWindowWMInfo() isn't filling in the egl_window handle nor the display handle in the Wayland case, but works fine when using x11.  I wrote a test program which just inits SDL2, creates a window and gets the handles in the exact same way, using the same flags as MAME, and it works fine.  Somehow it just doesn't work.

Below is the verbose output starting MAME with both Wayland (fails) and x11(works) printing the nwh and ndt as filled in through SDL_GetWindowWMInfo():
```
Attempting load of mame.ini
Parsing mame.ini
Available videodrivers: wayland x11 KMSDRM offscreen dummy evdev
Current Videodriver: wayland
	Display #0
		Renderdrivers:
			    opengl (0x0)
			 opengles2 (0x0)
			  opengles (0x0)
			  software (0x0)
Available audio drivers:
	pulseaudio
	sndio
	jack
	pipewire
	disk
	dummy
Build version:      0.255 (unknown)
Build architecure:
Build defines 1:    SDLMAME_UNIX=1 SDLMAME_LINUX=1
Build defines 1:    LSB_FIRST=1 PTR64=1
SDL/OpenGL defines: SDL_COMPILEDVERSION=4701 USE_OPENGL=1
Compiler defines A: __GNUC__=12 __GNUC_MINOR__=2 __GNUC_PATCHLEVEL__=1 __VERSION__="12.2.1 20230304"
Compiler defines B: __amd64__=1 __x86_64__=1 __unix__=1
Compiler defines C: __USE_FORTIFY_LEVEL=0
Enter init_monitors
Adding monitor screen0 (1920 x 1080)
Leave init_monitors
Enter sdlwindow_init

Hints:
	SDL_FRAMEBUFFER_ACCELERATION             (NULL)
	SDL_RENDER_DRIVER                        (NULL)
	SDL_RENDER_OPENGL_SHADERS                (NULL)
	SDL_RENDER_SCALE_QUALITY                 (NULL)
	SDL_RENDER_VSYNC                         (NULL)
	SDL_VIDEO_X11_XVIDMODE                   (NULL)
	SDL_VIDEO_X11_XINERAMA                   (NULL)
	SDL_VIDEO_X11_XRANDR                     (NULL)
	SDL_GRAB_KEYBOARD                        (NULL)
	SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS         (NULL)
	SDL_IOS_IDLE_TIMER_DISABLED              (NULL)
	SDL_IOS_ORIENTATIONS                     (NULL)
	SDL_XINPUT_ENABLED                       (NULL)
	SDL_GAMECONTROLLERCONFIG                 (NULL)
	SDL_JOYSTICK_ALLOW_BACKGROUND_EVENTS     (NULL)
	SDL_ALLOW_TOPMOST                        (NULL)
	SDL_TIMER_RESOLUTION                     (NULL)
	SDL_RENDER_DIRECT3D_THREADSAFE           (NULL)
	SDL_VIDEO_ALLOW_SCREENSAVER              (NULL)
	SDL_ACCELEROMETER_AS_JOYSTICK            (NULL)
	SDL_MAC_CTRL_CLICK_EMULATE_RIGHT_CLICK   (NULL)
	SDL_VIDEO_WIN_D3DCOMPILER                (NULL)
	SDL_VIDEO_WINDOW_SHARE_PIXEL_FORMAT      (NULL)
	SDL_VIDEO_MAC_FULLSCREEN_SPACES          (NULL)
	SDL_MOUSE_RELATIVE_MODE_WARP             (NULL)
	SDL_RENDER_DIRECT3D11_DEBUG              (NULL)
	SDL_VIDEO_HIGHDPI_DISABLED               (NULL)
	SDL_WINRT_PRIVACY_POLICY_URL             (NULL)
	SDL_WINRT_PRIVACY_POLICY_LABEL           (NULL)
	SDL_WINRT_HANDLE_BACK_BUTTON             (NULL)
Leave sdlwindow_init
3840x2160@60 -> 0.000205
3200x1800@60 -> 0.000258
2880x1620@60 -> 0.000296
2560x1600@60 -> 0.000329
2560x1440@60 -> 0.000347
2048x1536@60 -> 0.000406
2048x1152@60 -> 0.000481
1920x1440@60 -> 0.000446
1920x1200@60 -> 0.000500
1920x1080@60 -> 0.000532
1680x1050@60 -> 0.000621
1600x1200@60 -> 0.000595
1600x 900@60 -> 0.000724
1440x1080@60 -> 0.000714
1440x 900@60 -> 0.000819
1400x1050@60 -> 0.000751
1368x 768@60 -> 0.000983
1280x1024@60 -> 0.000844
1280x 960@60 -> 0.000892
1280x 800@60 -> 0.001041
1280x 720@60 -> 0.001135
1152x 864@60 -> 0.001115
1024x 768@60 -> 0.001486
 864x 486@60 -> 0.004329
 800x 600@60 -> 0.003559
 720x 480@60 -> 0.012346
 640x 480@60 -> 1.000000
Enter sdl_info::create
Initializing BGFX library
nwh=335663952, ndt=333654512   <--- here nwh doesn't seem like an egl_window handle
../../../../../3rdparty/bgfx/src/glcontext_egl.cpp (278): BGFX FATAL 0x00000002: Failed to create surface.

```
```
Attempting load of mame.ini
Parsing mame.ini
Available videodrivers: wayland x11 KMSDRM offscreen dummy evdev
Current Videodriver: x11
	Display #0
		Renderdrivers:
			    opengl (0x0)
			 opengles2 (0x0)
			  opengles (0x0)
			  software (0x0)
Available audio drivers:
	pulseaudio
	sndio
	jack
	pipewire
	disk
	dummy
Build version:      0.255 (unknown)
Build architecure:
Build defines 1:    SDLMAME_UNIX=1 SDLMAME_LINUX=1
Build defines 1:    LSB_FIRST=1 PTR64=1
SDL/OpenGL defines: SDL_COMPILEDVERSION=4701 USE_OPENGL=1
Compiler defines A: __GNUC__=12 __GNUC_MINOR__=2 __GNUC_PATCHLEVEL__=1 __VERSION__="12.2.1 20230304"
Compiler defines B: __amd64__=1 __x86_64__=1 __unix__=1
Compiler defines C: __USE_FORTIFY_LEVEL=0
Enter init_monitors
Adding monitor screen0 (3840 x 2160)
Leave init_monitors
Enter sdlwindow_init

Hints:
	SDL_FRAMEBUFFER_ACCELERATION             (NULL)
	SDL_RENDER_DRIVER                        (NULL)
	SDL_RENDER_OPENGL_SHADERS                (NULL)
	SDL_RENDER_SCALE_QUALITY                 (NULL)
	SDL_RENDER_VSYNC                         (NULL)
	SDL_VIDEO_X11_XVIDMODE                   (NULL)
	SDL_VIDEO_X11_XINERAMA                   (NULL)
	SDL_VIDEO_X11_XRANDR                     (NULL)
	SDL_GRAB_KEYBOARD                        (NULL)
	SDL_VIDEO_MINIMIZE_ON_FOCUS_LOSS         (NULL)
	SDL_IOS_IDLE_TIMER_DISABLED              (NULL)
	SDL_IOS_ORIENTATIONS                     (NULL)
	SDL_XINPUT_ENABLED                       (NULL)
	SDL_GAMECONTROLLERCONFIG                 (NULL)
	SDL_JOYSTICK_ALLOW_BACKGROUND_EVENTS     (NULL)
	SDL_ALLOW_TOPMOST                        (NULL)
	SDL_TIMER_RESOLUTION                     (NULL)
	SDL_RENDER_DIRECT3D_THREADSAFE           (NULL)
	SDL_VIDEO_ALLOW_SCREENSAVER              (NULL)
	SDL_ACCELEROMETER_AS_JOYSTICK            (NULL)
	SDL_MAC_CTRL_CLICK_EMULATE_RIGHT_CLICK   (NULL)
	SDL_VIDEO_WIN_D3DCOMPILER                (NULL)
	SDL_VIDEO_WINDOW_SHARE_PIXEL_FORMAT      (NULL)
	SDL_VIDEO_MAC_FULLSCREEN_SPACES          (NULL)
	SDL_MOUSE_RELATIVE_MODE_WARP             (NULL)
	SDL_RENDER_DIRECT3D11_DEBUG              (NULL)
	SDL_VIDEO_HIGHDPI_DISABLED               (NULL)
	SDL_WINRT_PRIVACY_POLICY_URL             (NULL)
	SDL_WINRT_PRIVACY_POLICY_LABEL           (NULL)
	SDL_WINRT_HANDLE_BACK_BUTTON             (NULL)
Leave sdlwindow_init
3840x2160@60 -> 0.000205
3200x1800@60 -> 0.000258
2880x1620@60 -> 0.000296
2560x1600@60 -> 0.000329
2560x1440@60 -> 0.000347
2048x1536@60 -> 0.000406
2048x1152@60 -> 0.000481
1920x1440@60 -> 0.000446
1920x1200@60 -> 0.000500
1920x1080@60 -> 0.000532
1680x1050@60 -> 0.000621
1600x1200@60 -> 0.000595
1600x 900@60 -> 0.000724
1440x1080@60 -> 0.000714
1440x 900@60 -> 0.000819
1400x1050@60 -> 0.000751
1368x 768@60 -> 0.000983
1280x1024@60 -> 0.000844
1280x 960@60 -> 0.000892
1280x 800@60 -> 0.001041
1280x 720@60 -> 0.001135
1152x 864@60 -> 0.001115
1152x 720@60 -> 0.001328
1024x 768@60 -> 0.001486
1024x 576@60 -> 0.002079
 960x 600@60 -> 0.002268
 928x 580@60 -> 0.002571
 864x 486@60 -> 0.004329
 800x 600@60 -> 0.003559
 800x 500@59 -> 0.005525
 768x 480@60 -> 0.007752
 720x 480@60 -> 0.012346
 720x 400@60 -> 0.000006
 640x 480@59 -> 1.000000
 640x 400@60 -> 0.000012
 640x 350@60 -> 0.000008
 320x 240@60 -> 0.000002
 320x 200@59 -> 0.000002
Enter sdl_info::create
Initializing BGFX library
nwh=4194306, ndt=333658448    <-- here nwh is a I would expect from my test code.
Audio: Start initialization
Audio: Driver is pulseaudio
Audio: frequency: 48000, channels: 2, samples: 512
sdl_create_buffers: creating stream buffer of 25600 bytes
Audio: End initialization
...
```

SDL_GetWindowWMInfo() called from:
https://github.com/mamedev/mame/blob/master/src/osd/modules/render/drawbgfx.cpp

Window created here:
https://github.com/mamedev/mame/blob/master/src/osd/sdl/window.cpp

In both cases I "ndt" seems wrong, at least compared to my test code:

```
./sdltest
Native Window Handle (x11) is 4194312, ndt is 4315616
Native Window Handle (wayland) is 6424224, ndt is 4307904
```

**Comment by Kontrabant** [comment]
The bug is in BGFX, which crashes on Wayland, because it doesn't check the window type and treats everything on Linux as X11.  If you backtrace, it's crashing in `XGetXCBConnection`, because it's aliasing the Wayland `wl_display` handle to an X11 `Display` handle.  See also https://github.com/bkaradzic/bgfx/issues/2416

On top of that, even if BGFX had a proper Wayland path that handled things correctly, MAME passing the EGL window handle to BGFX is wrong, as if you are using Vulkan, the `VkWaylandSurfaceCreateInfoKHR` struct needs the `wl_surface`, not the EGL window.

Closing as not our bug.

**Comment by sjnewbury** [comment]
For what it's worth, I did build MAME with the below patch which *did* make BGFX use GLES3.2, still working with XWayland despite NO_X11.

It still doesn't explain why I wasn't getting a correct nwh, even if inserted the code immediately after window creation and it was the same.  I suspect it is something wrong in MAME since, as I said, a simple test example works fine. It was really confusing me.

I did wonder how Vulkan is supposed to work!
```
--- ./scripts/src/3rdparty.lua~ 2023-05-30 18:50:57.000000000 +0100
+++ ./scripts/src/3rdparty.lua  2023-06-11 23:49:06.161173772 +0100
@@ -1457,8 +1457,12 @@
        if _OPTIONS["targetos"]=="linux" or _OPTIONS["targetos"]=="netbsd" or _OPTIONS["targetos"]=="openbsd" then
                if _OPTIONS["NO_X11"]=="1" then
                        defines {
-                               "BGFX_CONFIG_RENDERER_OPENGLES=1",
+                               "BGFX_CONFIG_RENDERER_OPENGLES=32",
                                "BGFX_CONFIG_RENDERER_OPENGL=0",
+                               "WL_EGL_PLATFORM=1",
+                       }
+

```

**Comment by Kontrabant** [comment]
I'm still not sure what's 'incorrect' about the values you are seeing.  SDL is populating the struct correctly, and other applications which use these values work just fine.  Bear in mind that MAME is passing a `Window` on X11 and an `wl_egl_window` for Wayland, so the handles are pointers to completely different resources.

Either way, the root cause here is the lack of native Wayland support in BGFX, so there's nothing for SDL to do.

**Comment by sjnewbury** [comment]
I expected the values to look the similar to the test code I wrote which simply initialises SDL, creates a window, and then prints the relevant info.

nwh=335663952 (from MAME) vs Native Window Handle (wayland) is 6424224 (my test program)

It's consistent and reproducible, I found it suspicious.  I wondered whether it was actually creating a Wayland window at all!

Supposedly, BGFX does have some Wayland support, when built with "--with-wayland", but having read the bug report you linked above I do wonder whether it ever actually worked.  I'll try taking a look at the latest upstream BGFX itself, and trying the examples.
